mini-projects/01-device-driver/laptop-interface/src/controller.rs

```diff
@@ -0,0 +1,516 @@
+#![allow(dead_code)]
+
+use crate::error::{ControllerError, ModuleError, SerialError};
+use crate::lib_serial_ffi::*;
+
+use std::ffi::CString;
+use std::sync::mpsc::{Receiver, Sender, TryRecvError};
+use std::thread::{self, JoinHandle};
+use std::time::Duration;
+use std::{default, str};
+
+use log::{self, error, info, trace};
+use regex::Regex;
+use std::collections::HashMap;
+
+pub struct ControllerManager {
```

Comment:
I really like how clean this interface turned out, creates a really nice separation of logic between each controller component